src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -1107,9 +1114,48 @@ private bool ShouldExecuteCodeBlockActions(AnalysisScope analysisScope, ISymbol
                 return computeDeclarationAnalysisData();
             }
 
-            return _declarationAnalysisDataCache.GetValue(declaration, decl => computeDeclarationAnalysisData());
+            var map = _declarationAnalysisDataCache.GetValue(compilation, _ => new Dictionary<Location, DeclarationAnalysisData>());
+            var location = declaration.GetLocation();
+
+            DeclarationAnalysisData data;
+            lock (_gate)
+            {
+                if (map.TryGetValue(location, out data))
+                {
+                    return data;
+                }
+            }
+
+            data = computeDeclarationAnalysisData();
+
+            lock (_gate)
```

Comment:
The driver guarantees that only a single thread will be performing analysis on individual declaration. However, there might be multiple threads analyzing different trees at the same time, and hence the need for the lock.